xarray/core/coordinates.py

```diff
@@ -23,49 +34,58 @@
 _THIS_ARRAY = ReprObject("<this-array>")
 
 
-class AbstractCoordinates(collections.abc.Mapping):
-    def __getitem__(self, key):
-        raise NotImplementedError
+class AbstractCoordinates(Mapping[Hashable, "DataArray"]):
+    _data = None  # type: Union["DataArray", "Dataset"]
 
-    def __setitem__(self, key, value):
+    def __getitem__(self, key: Hashable) -> "DataArray":
+        raise NotImplementedError()
+
+    def __setitem__(self, key: Hashable, value: Any) -> None:
         self.update({key: value})
 
     @property
-    def indexes(self):
+    def _names(self) -> Set[Hashable]:
+        raise NotImplementedError()
+
+    @property
+    def dims(self) -> Union[Mapping[Hashable, int], Tuple[Hashable, ...]]:
```

Comment:
Why the change here from `return self._data.dims`?

Reply:
Because ``DataArray.dims`` is a ``Tuple[Hashable, ...]`` whereas ``Dataset.dims`` is a ``Mapping[Hashable, int]``

Reply:
See below for the two implementations, which are identical in code but different in signature.
The method needs stay in the abstract class because other methods of the abstract class invoke it; this signature with NotImplementedError() guarantees that the methods of the abstract class can only use dims as the common denominator of the two output types, which is a ``Collection[Hashable]``.

Reply:
OK great